src/misc/placable-items.ts

```diff
@@ -0,0 +1,219 @@
+import { GameMode, Location, Material, SoundCategory } from 'org.bukkit';
+import { Block, BlockFace } from 'org.bukkit.block';
+import { ArmorStand, EntityType, Player } from 'org.bukkit.entity';
+import { Action } from 'org.bukkit.event.block';
+import {
+  PlayerInteractAtEntityEvent,
+  PlayerInteractEvent,
+  PlayerArmorStandManipulateEvent,
+} from 'org.bukkit.event.player';
+import { EquipmentSlot, ItemStack } from 'org.bukkit.inventory';
+import { giveItem } from '../common/helpers/inventory';
+import { EulerAngle } from 'org.bukkit.util';
+
+/**
+ * Allow specific items to be placed on the table as decorations.
+ * Items like: bottles, bowls, books
+ */
+
+const PLACABLE_MATERIALS = new Set([
+  Material.GLASS_BOTTLE,
+  Material.POTION,
+  Material.BOOK,
+  Material.WRITABLE_BOOK,
+  Material.WRITTEN_BOOK,
+
+  Material.BOWL,
+  Material.BEETROOT_SOUP,
+  Material.SUSPICIOUS_STEW,
+  Material.MUSHROOM_STEW,
+  Material.RABBIT_STEW,
+]);
+
+const TABLES = new Set([
+  Material.OAK_PLANKS,
+  Material.SPRUCE_PLANKS,
+  Material.BIRCH_PLANKS,
+  Material.JUNGLE_PLANKS,
+  Material.ACACIA_PLANKS,
+  Material.DARK_OAK_PLANKS,
+
+  Material.OAK_SLAB,
+  Material.SPRUCE_SLAB,
+  Material.BIRCH_SLAB,
+  Material.JUNGLE_SLAB,
+  Material.ACACIA_SLAB,
+  Material.DARK_OAK_SLAB,
+
+  Material.OAK_STAIRS,
+  Material.SPRUCE_STAIRS,
+  Material.BIRCH_STAIRS,
+  Material.JUNGLE_STAIRS,
+  Material.ACACIA_STAIRS,
+  Material.DARK_OAK_STAIRS,
+]);
+
+/**
+ * Check if the item can be placed on the ground
+ */
+function isPlacableItem(item: ItemStack | null): item is ItemStack {
+  if (!item) return false;
+  return PLACABLE_MATERIALS.has(item.type);
+}
+
+/**
+ * Check if items can be placed on this block
+ */
+function isValidSurface(block: Block | null): block is Block {
+  if (!block) return false;
+  return TABLES.has(block.type);
+}
+
+/**
+ * Place item on the ground
+ */
+registerEvent(PlayerInteractEvent, (event) => {
+  if (event.action !== Action.RIGHT_CLICK_BLOCK) return;
+  if (event.blockFace !== BlockFace.UP) return;
+  if (!isValidSurface(event.clickedBlock)) return;
+  if (!isPlacableItem(event.item)) return;
+
+  const location = getClickedLocation(event.player);
+  if (!location) return;
+
+  if (containsTooManyItems(location)) {
+    warnTooManyItems(event.player);
+    return;
+  }
+
+  spawnHolderArmorStand(location, event.item);
+  if (event.player.gameMode !== GameMode.CREATIVE) event.item.amount--;
+  event.setCancelled(true);
```

Comment:
Move canceling the event before too many items warning. Chances are that players expect *nothing* to happen when they see it.